### PR TITLE
LUCENE-10458: BoundedDocSetIdIterator may supply error count in Weigth#count(LeafReaderContext) when missingValue enables

### PR DESCRIPTION
**Pull request by LuXugang**
When IndexSortSortedNumericDocValuesRangeQuery can take advantage of index sort, Weight#count will use BoundedDocSetIdIterator's lastDoc and firstDoc to calculate count, but if missingValue enables, those Documents which not contain DocValues may be involved in calculating count.

That is why in BoundedDocSetIdIterator#advance(int target), target need to be check if such Document contains DocValues
`int result = delegate.advance(target);`

### CODE REVIEW

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 1]
```diff
@@ -198,16 +198,22 @@ public boolean isCacheable(LeafReaderContext ctx) {
...
       @Override
       public int count(LeafReaderContext context) throws IOException {
-        BoundedDocSetIdIterator disi = getDocIdSetIteratorOrNull(context);
-        if (disi != null) {
-          return disi.lastDoc - disi.firstDoc;
+        Sort indexSort = context.reader().getMetaData().getSort();
+        if (indexSort != null
+            && indexSort.getSort().length > 0
+            && indexSort.getSort()[0].getField().equals(field)
+            && indexSort.getSort()[0].getMissingValue() == null) {
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:205]: I don't think that this is the right thing to check since the missing value is assumed to be zero if not set.

The best thing we can do that I can think of is to check if the field is dense via points (ie. no missing values) or if the missing value falls outside of the range so that the bounded iterator is accurate?

**Review by jpountz — CHANGES_REQUESTED** [review]
Thanks for catching this!

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java (hunk 1 above)
Comment (LuXugang) [reply lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:205]: `indexSort.getSort()[0].getMissingValue() == null`
It indeed seems too aggressive, Thanks.

**Review by jpountz — COMMENTED** [review]
I wonder if we could simplify this and remove the lambda.

One idea I'm contemplating is to merge this with your idea on LUCENE-10460. Maybe we could create the `BoundedDocSetIdIterator` with a `null` `delegate` when all docs within the range are known to match. Then the scorer would check the delegate when it's not null, and `count` would only trust the count if `delegate` is `null`?

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 2]
```diff
@@ -380,4 +372,25 @@ public long cost() {
...
+      }
+
+      if (target < lastDoc) {
+        docID = target;
+      } else {
+        docID = NO_MORE_DOCS;
+      }
+      return docID;
+    }
+  }
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:395]: nit: I'd rather fold this logic into `BoundedDocIdSetIterator` with a `null` check on `delegate` unless we have evidence that the sub class buys us much?

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 3]
```diff
@@ -307,9 +292,16 @@ private BoundedDocIdSetIterator getDocIdSetIterator(
...
 
     int lastDocIdExclusive = high + 1;
-    BoundedDocIdSetIterator disi =
-        new BoundedDocIdSetIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
-    return boundedDisiFunction.apply(sortField.getMissingValue(), disi);
+    Object missingValue = sortField.getMissingValue();
+    BoundedDocIdSetIterator disi;
+    if (missingValue == null
+        || (long) missingValue < lowerValue
+        || (long) missingValue > upperValue) {
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:299]: I don't think it's correct to exclude the `missingValue == null` case since it's assumed to be zero by default? We should resolve the missing value that is used in practice first, e.g. as below?

```
final long missingLongValue = missingValue == null ? 0L : (long) missingValue;
if (missingLongValue < lowerValue || missingValue > upperValue) { ... }
```

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 4]
```diff
@@ -201,27 +200,18 @@ public boolean isCacheable(LeafReaderContext ctx) {
...
-        }
-
+        LeafReader reader = context.reader();
+        PointValues pointValues = reader.getPointValues(field);
+        BoundedDocIdSetIterator disi = getDocIdSetIteratorOrNull(context);
         if (disi != null) {
-          return disi.lastDoc - disi.firstDoc;
+          if (reader.hasDeletions() == false
+                  && pointValues != null
+                  && pointValues.getDocCount() == context.reader().maxDoc()
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:209]: this `docCount == maxDoc` check could move to `getDocIdSetIterator` to set the `delegate` to `null` when the field is dense?

**Review by jpountz — COMMENTED** [review]
Thanks, this looks almost ready!

Can you move the CHANGES entry from `LUCENE-10385` to the 9.2 section and add your name to the list of authors?

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 5]
```diff
@@ -278,7 +287,21 @@ private BoundedDocSetIdIterator getDocIdSetIterator(
...
+    Object missingValue = sortField.getMissingValue();
+    BoundedDocIdSetIterator disi;
+    LeafReader reader = context.reader();
+    PointValues pointValues = reader.getPointValues(field);
+    final long missingLongValue = missingValue == null ? 0L : (long) missingValue;
+    if (reader.hasDeletions() == false
+            && pointValues != null
+            && pointValues.getDocCount() == reader.maxDoc()
+        || missingLongValue < lowerValue
+        || missingLongValue > upperValue) {
```
Comment (iverase) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:299]: This if condition is extremely difficult to read, I wonder if we should add some parenthesis or rewrite it somehow to make it clear

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 6]
```diff
@@ -278,7 +287,21 @@ private BoundedDocSetIdIterator getDocIdSetIterator(
     }
 
     int lastDocIdExclusive = high + 1;
-    return new BoundedDocSetIdIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
+    Object missingValue = sortField.getMissingValue();
```
Comment (jtibshirani) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:290]: It feels surprising to be checking deletions here, since it's not relevant to the query case -- maybe we could pull that into the `count` method.

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 7]
```diff
@@ -459,54 +461,55 @@ public void testIndexSortOptimizationDeactivated(RandomIndexWriter writer) throw
     reader.close();
   }
 
-  public void testCount() throws IOException {
+  public void testCountHasMissingValue() throws IOException {
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:464]: It could be really helpful to add another test like `testSameHitsAsPointRangeQuery` that compares the counts of the two queries. Also in `testSameHitsAsPointRangeQuery` I noticed that we always set `sortField.setMissingValue(random().nextLong())`, it'd be great to randomize this more -- sometimes not setting it at all, sometimes setting it to `Long.MIN_VALUE` or `Long.MAX_VALUE`.

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 8]
```diff
@@ -198,16 +200,23 @@ public boolean isCacheable(LeafReaderContext ctx) {
...
-        BoundedDocSetIdIterator disi = getDocIdSetIteratorOrNull(context);
-        if (disi != null) {
+
+        BoundedDocIdSetIterator disi = getDocIdSetIteratorOrNull(context);
+        if (disi != null && disi.delegate == null) {
           return disi.lastDoc - disi.firstDoc;
         }
-        return fallbackWeight.count(context);
+
+        PointValues pointValues = context.reader().getPointValues(field);
```
Comment (jtibshirani) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:209]: If there's a fallback query provided, it's guaranteed to match the same documents (from the javadoc: "This fallback must be an equivalent range query -- it should produce the same documents and give constant scores"). There's nothing in this query that guarantees the fallback query uses point values, so the check feels out of place.

**Review by jtibshirani — COMMENTED** [review]
Sorry for jumping in late, I just saw this PR. I am a little worried about the increasing complexity in this query -- I always have to stare at it a long time to remember how everything works! But I understand this optimization can make a difference in the count case.

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 9]
```diff
@@ -306,14 +329,14 @@ private static ValueComparator loadComparator(
...
-  private static class BoundedDocSetIdIterator extends DocIdSetIterator {
-    private final int firstDoc;
-    private final int lastDoc;
+  private static class BoundedDocIdSetIterator extends DocIdSetIterator {
+    protected final int firstDoc;
+    protected final int lastDoc;
     private final DocIdSetIterator delegate;
 
-    private int docID = -1;
+    protected int docID = -1;
```
Comment (jtibshirani) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:337]: I think this could still be private.

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 10]
```diff
@@ -59,7 +61,14 @@ public void testSameHitsAsPointRangeQuery() throws IOException {
       IndexWriterConfig iwc = new IndexWriterConfig(new MockAnalyzer(random()));
       boolean reverse = random().nextBoolean();
       SortField sortField = new SortedNumericSortField("dv", SortField.Type.LONG, reverse);
-      sortField.setMissingValue(random().nextLong());
+      boolean enableMissingValue = random().nextBoolean();
+      if (enableMissingValue) {
+        long missingValue =
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:66]: I find this check a little hard to follow and the probabilities end up being surprising -- 1/2 for MIN_VALUE, 1/4 for MAX_VALUE, 1/4 a random long. Maybe we could do half the time a random long, half the time randomly choose between MIN_VALUE and MAX_VALUE.

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 11]
```diff
@@ -459,56 +468,122 @@ public void testIndexSortOptimizationDeactivated(RandomIndexWriter writer) throw
...
 
-  public void testFallbackCount() throws IOException {
+  public void testCountBoundary() throws IOException {
     Directory dir = newDirectory();
     IndexWriterConfig iwc = new IndexWriterConfig(new MockAnalyzer(random()));
-    Sort indexSort = new Sort(new SortedNumericSortField("field", SortField.Type.LONG));
+    SortField sortField = new SortedNumericSortField("field", SortField.Type.LONG);
+    // true: lower
+    // false: upper
+    boolean lowerOrUpper = random().nextBoolean();
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:542]: Small comment, maybe `useLower` would be clearer here, then you wouldn't need the comment explanation.

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 12]
```diff
@@ -459,56 +468,122 @@ public void testIndexSortOptimizationDeactivated(RandomIndexWriter writer) throw
...
-    reader.close();
-    dir.close();
+  private void assertSameCount(Weight weight1, Weight weight2, IndexSearcher searcher)
+      throws IOException {
+    for (LeafReaderContext context : searcher.getLeafContexts()) {
+      assertEquals(weight1.count(context), weight2.count(context));
+    }
   }
 
-  public void testFallbackCount() throws IOException {
```
Comment (jtibshirani) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:486]: Could you explain why you removed this test? It seemed helpful to me.

File: lucene/CHANGES.txt [hunk 13]
```diff
@@ -56,14 +56,17 @@ API Changes
 
 New Features
 ---------------------
-(No changes)
+
+* LUCENE-10385: Implement Weight#count on IndexSortSortedNumericDocValuesRangeQuery
+  to speed up computing the number of hits when possible. (Lu Xugang, Cavanna, Adrien Grand)
```
Comment (jtibshirani) [inline lucene/CHANGES.txt:61]: We could keep Luca's name here: Luca Cavanna :)

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java (hunk 6 above)
Comment (jtibshirani) [reply lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:283]: Pinging again on this: what do you think of moving `reader.hasDeletions` up to the check inside `count` instead?

**Review by jtibshirani — COMMENTED** [review]
Thank you for all the iterations! It looks good to me overall, I just left a couple small comments.

**Review by jtibshirani — COMMENTED** [review]
This looks good to me except the one comment that hasn't been addressed yet about moving `reader.hasDeletions()`. Also maybe @jpountz would like to take another look.

File: lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java [hunk 14]
```diff
@@ -278,7 +280,21 @@ private BoundedDocSetIdIterator getDocIdSetIterator(
...
 
     int lastDocIdExclusive = high + 1;
-    return new BoundedDocSetIdIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
+    Object missingValue = sortField.getMissingValue();
+    BoundedDocIdSetIterator disi;
+    LeafReader reader = context.reader();
+    PointValues pointValues = reader.getPointValues(field);
+    final long missingLongValue = missingValue == null ? 0L : (long) missingValue;
+    // all documents have docValues or missing value falls outside the range
+    if ((reader.hasDeletions() == false
```
Comment (jpountz) [inline lucene/sandbox/src/java/org/apache/lucene/sandbox/search/IndexSortSortedNumericDocValuesRangeQuery.java:289]: Can you mave this `hasDeletions` check inside `count` instead as @jtibshirani suggested? It would still be ok to set the delegate to `null` for `scorer` if there are deletions since deletes are applied on top of the `Scorer` API.

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 15]
```diff
@@ -59,7 +61,14 @@ public void testSameHitsAsPointRangeQuery() throws IOException {
       IndexWriterConfig iwc = new IndexWriterConfig(new MockAnalyzer(random()));
       boolean reverse = random().nextBoolean();
       SortField sortField = new SortedNumericSortField("dv", SortField.Type.LONG, reverse);
-      sortField.setMissingValue(random().nextLong());
+      boolean enableMissingValue = random().nextBoolean();
+      if (enableMissingValue) {
+        long missingValue =
+            random().nextBoolean()
+                ? random().nextLong()
```
Comment (jpountz) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:68]: The problem with a random long is that almost all the time it would fall outside of the query ranges that we test below, so we would have no coverage for the case when the missing value is set and falls within the query range.

Can you use a random long between -100 and 10,000 like other values or something like that instead?

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 16]
```diff
@@ -471,44 +480,134 @@ public void testCount() throws IOException {
...
+    final int iters = atLeast(10);
+    for (int iter = 0; iter < iters; ++iter) {
+      Directory dir = newDirectory();
+      IndexWriterConfig iwc = new IndexWriterConfig(new MockAnalyzer(random()));
+      SortField sortField = new SortedNumericSortField("field", SortField.Type.LONG);
+      boolean enableMissingValue = random().nextBoolean();
+      if (enableMissingValue) {
+        long missingValue =
+            random().nextBoolean()
+                ? random().nextLong()
```
Comment (jpountz) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:507]: and likewise here

File: lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java [hunk 17]
```diff
@@ -471,44 +480,134 @@ public void testCount() throws IOException {
...
+
+      final int numDocs = atLeast(100);
+      for (int i = 0; i < numDocs; ++i) {
+        Document doc = new Document();
+        final int numValues = TestUtil.nextInt(random(), 0, 1);
+        for (int j = 0; j < numValues; ++j) {
+          final long value = TestUtil.nextLong(random(), -100, 10000);
+          doc =
+              random().nextBoolean()
+                  ? createMissingValueDocument()
```
Comment (jpountz) [inline lucene/sandbox/src/test/org/apache/lucene/sandbox/search/TestIndexSortSortedNumericDocValuesRangeQuery.java:521]: Hmm, I don't understand this, are we not already creating a document that has a missing value when `numValues` is equal to 0?

**Review by jpountz — CHANGES_REQUESTED** [review]
I left a few more comments, mostly about tests, but this looks almost ready to go to me.

**Review by jpountz — APPROVED** [review]
